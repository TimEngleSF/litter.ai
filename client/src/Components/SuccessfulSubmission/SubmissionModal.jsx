import React from 'react';
import parse from 'html-react-parser';

export default function SubmissionModal({
  modalOpen,
  setModalOpen,
  type,
}) {
  return (
    <div className="modal-container" data-testid="modal">
      <div className="modal-container-top">
        <h2>{type.name}</h2>
        <button
          type="button"
          className="modal-x-button"
          onClick={() => { setModalOpen(!modalOpen); }}
        >
          &#x2715;
        </button>
      </div>
      <hr />
      <div className="modal-text">{parse(type.desc_long)}</div>
      <button
        type="button"
        className="modal-got-it-button"
        data-testid="modal-got-it-button"
        onClick={() => { setModalOpen(!modalOpen); }}
      >
        Got it
      </button>
    </div>
  );
}
