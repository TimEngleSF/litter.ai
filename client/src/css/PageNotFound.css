.PNF-container {
  height: 100dvh;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.PNF-container
.PNF-404 {
  display: flex;
  justify-content: center;
  align-items: center;
  position: relative;
}

.PNF-container
.PNF-404
.PNF-icon {
  font-size: 10rem;
  position: absolute;
  color: var(--main-green)
}

.PNF-container
.PNF-404 h1 {
  font-size: 11rem;
  margin: 0;
}

.PNF-container h3 {
  margin: -1rem 0 5rem 0;
}

.PNF-container
.PNF-buttons button {
  width: 15dvw;
}

/* Phones */
@media only screen and (max-width: 600px) {

}

/* Portrait tablets and large phones */
@media only screen and (min-width: 601px) and (max-width: 1024px) {

}

/* Small laptops and desktops */
@media only screen and (min-width: 1025px) {

}