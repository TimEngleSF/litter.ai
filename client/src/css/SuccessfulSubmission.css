.category-container {
  width: 100%;
  height: 100dvh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.category-container
.category-wrapper {
  height: 100dvh;
  width: 100dvw;
  max-width: 768px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
}

.category-container
.category-wrapper h4 {
  font-size: 1.5rem;
  padding-top: 4rem;
  margin-bottom: 0;
}

.category-container
.category-wrapper h1 {
  text-transform: capitalize;
}

.category-container
.category-wrapper
.category-icon {
  font-size: 10rem;
  color: var(--main-green);
}

.category-container
.category-wrapper
.category-short-desc {
  width: 85%;
}

.category-container
.category-wrapper
.category-short-desc p {
  font-size: 1.2rem;
  text-align: left;
  margin: 0.7rem 0;
}

.category-container
.category-modal {
  width: 35dvw;
  max-width: 100%;
  max-height: 45dvh;
  display: flex;
  justify-content: center;
  align-items: center;
  position: absolute;
  z-index: 2;
}

/******** Modal ********/

.modal-container {
  min-height: 20dvh;
  max-height: 60dvh;
  background-color: var(--white);
  box-shadow: 0.75em 0.75em 1.75em -0.5em var(--shadow-color);
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  border-radius: 1.5rem;
  z-index: 2;
}

.modal-container
.modal-container-top {
  width: 85%;
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
}

.modal-container
.modal-container-top h2 {
  margin: 2rem 0 0 0;
  text-transform: capitalize;
}

.modal-container
.modal-x-button {
  border: none;
  background: none;
  font-size: 1rem;
  font-weight: bold;
  color: var(--black);
}

.modal-container
.modal-x-button:hover {
  cursor: pointer;
}

.modal-container hr {
  border: none;
  height: 1px;
  background-color: #000;
  margin-top: 1rem;
  width: 90%;
}

.modal-container p {
  margin: 0.5rem 0;
}

.modal-container
.modal-text {
  width: 85%;
  max-width: 90%;
  height: 100%;
  margin: 0;
  overflow-y: auto;
  padding: 0 0.6rem;
}

.modal-container
.modal-text h1 {
  margin: 0.5rem 0 0 0;
  font-size: 1.2rem;
}

.modal-container
.modal-text ul {
  margin: 0.2rem 0 0 0;
  padding-left: 1.5rem;
}

.modal-container
.modal-text li {
  font-size: 1.3rem;
}


.modal-container
.modal-text::-webkit-scrollbar-thumb,
.modal-text::-webkit-scrollbar-track {
  background-color: transparent;
}

.modal-container
.modal-got-it-button {
  width: 50%;
  height: 7dvh;
  border-radius: 10rem;
  border: none;
  color: var(--white);
  background-color: var(--main-green);
  font-weight: bold;
  font-size: 1rem;
  margin: 1rem 0;
}

.modal-container
.modal-got-it-button:hover {
  cursor: pointer;
}

.modal-background {
  width: 100dvw;
  max-width: 100%;
  height: 100dvh;
  max-height: 100%;
  background-color: #000;
  background-color: #000000a3;
  position: absolute;
}


/* Phones */
@media only screen and (max-width: 600px) {
  .category-container
  .category-modal {
    width: 80dvw;
    max-height: 50dvh;
  }

  .category-container
  .category-wrapper h1 {
    font-size: 2rem;
    margin-bottom: 0;
  }

  .category-container
  .category-wrapper h4 {
    font-size: 1.3rem;
  }

  .category-container
  .category-wrapper
  .category-short-desc p {
    font-size: 1rem;
  }

  .category-container
  .category-wrapper
  .category-icon {
    font-size: 8rem;
  }

  .category-container
  .category-wrapper
  .category-buttons button {
    width: 85dvw;
  }

  .modal-container {
    height: 50dvh;
  }

  .modal-container
  .modal-got-it-button {
    height: 8dvh;
  }

  .modal-container
  .modal-text li {
    font-size: 0.9rem;
  }
}

/* Portrait tablets and large phones */
@media only screen and (min-width: 601px) and (max-width: 1024px) {
  .category-container
  .category-modal {
    width: 80dvw;
    max-height: 50dvh;
  }

  .category-container
  .category-wrapper h1 {
    font-size: 3.5rem;
    margin: 0;
  }

  .category-container
  .category-wrapper h4 {
    font-size: 2.5rem;
  }

  .category-container
  .category-wrapper
  .category-short-desc p {
    font-size: 1.8rem;
  }

  .category-container
  .category-wrapper
  .category-icon {
    font-size: 14rem;
  }

  .modal-container p {
    font-size: 1.5rem;
  }

  .modal-container
  .modal-container-top h2 {
    font-size: 2rem;
  }

  .modal-container
  .modal-x-button {
    font-size: 1.5rem;
  }

  .modal-container
  .modal-got-it-button{
    height: 7dvh;
    font-size: 1.5rem;
  }

  .category-container
  .category-wrapper
  .category-buttons
  button {
    font-size: 1.7rem;
    width: 85dvw;
  }
}

/* Small laptops and desktops */
@media only screen and (min-width: 1025px) {

}