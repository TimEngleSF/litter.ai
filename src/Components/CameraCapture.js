import React from 'react';


function CameraCapture () {
    return (
        <div>
            <input
            type="file" 
            accept=".jpg, .jpeg, .img, .HEIC"
            id="cameraFileInput"
            capture="environment"
            />
            <img id="pictureFromCamera" />

            <script src="script.js"></script>
        </div>
    )
}

export default CameraCapture;