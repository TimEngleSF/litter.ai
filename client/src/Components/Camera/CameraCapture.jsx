import React, { useState } from 'react';
import ImageUploaded from './ImageUploaded';
import AddImage from './AddImage';
import '../../css/Capture.css';

export default function CameraCapture() {
  const [image, setImage] = useState(null);

  return (
    <div className="capture-container main-container">
      {image
        ? <ImageUploaded image={image} setImage={setImage} />
        : <AddImage setImage={setImage} />}
      <script src="ImageCapture.js" />
    </div>
  );
}
