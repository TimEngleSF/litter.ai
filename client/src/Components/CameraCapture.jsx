import React, { useState, useRef } from 'react';
import { Link } from 'react-router-dom';
import { CiCirclePlus } from 'react-icons/ci';
import Dropdown from './Dropdown';
import '../css/Capture.css';

export default function CameraCapture() {
  const [image, setImage] = useState(null);
  const [imageSubmitted, setImageSubmitted] = useState(false);
  const [imageCategory, setImageCategory] = useState('glass');
  const captureRef = useRef(null);

  const openCamera = () => {
    captureRef.current.click();
  };

  const removePhoto = () => {
    setImage(null);
  };

  const handleImage = (e) => {
    const imageData = [e.target.files[0]];
    setImage({
      imagePreview: URL.createObjectURL(new Blob(imageData, { type: 'application/zip' })),
      imageFile: e.target.files[0],
    });
  };

  const handlePhotoSubmit = () => {
    setImageSubmitted(true);
  };

  return (
    <div className="capture-container main-container">
      {/* Image has been uploaded; preview the image */}
      {image
        ? (
          <div className="capture-image-wrapper">
            <div className="capture-image-preview">
              <button
                type="button"
                className="image-x-button"
                onClick={removePhoto}
              >
                &#x2715;
              </button>
              <img alt="" src={image.imagePreview} className="capture-image" />
            </div>
            {/* Image has been uploaded and submitted to the AI */}
            {imageSubmitted
              ? (
                <div className="capture-classify">
                  <h1>Help us classify</h1>
                  <div className="capture-dropdown">
                    <Dropdown
                      setImageCategory={setImageCategory}
                      imageCategory={imageCategory}
                    />
                  </div>
                  <p>
                    Is this&nbsp;
                    {imageCategory}
                    ?
                    If it is not,
                    please select the correct category from the dropdown.
                  </p>
                  <div className="capture-confirm-button lower-buttons">
                    <Link to={`/success/${imageCategory}`}>
                      <button type="button">
                        Confirm
                      </button>
                    </Link>
                  </div>
                </div>
              )
              : (
                <div className="capture-submit-button lower-buttons">
                  <button
                    type="button"
                    onClick={handlePhotoSubmit}
                  >
                    Submit
                  </button>
                </div>
              )}
          </div>
        )
        /* Image has not been uploaded, opens device's camera or gives option to upload an image */
        : (
          <>
            <div className="capture-top-section">
              <CiCirclePlus onClick={openCamera} className="capture-icon" />
              <input
                type="file"
                accept="image/*"
                id="cameraFileInput"
                ref={captureRef}
                capture="environment"
                onChange={handleImage}
                style={{ display: 'none' }}
              />
              <div className="capture-icon-bg" />
              <h1>Take a photo of the item</h1>
              <p>
                <span>Click on the camera to open up your device&#x27;s camera.</span>
                <br />
                <span>Please take the photo in portrait mode for best results.</span>
              </p>
            </div>
            <div className="capture-buttons lower-buttons">
              <Link to="/">
                <button className="button-home" type="button">
                  Home
                </button>
              </Link>
            </div>
          </>
        )}
      <script src="ImageCapture.js" />
    </div>
  );
}
