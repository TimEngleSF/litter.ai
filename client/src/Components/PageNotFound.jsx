import React from 'react';
import { Link } from 'react-router-dom';
import { MdOutlineWrongLocation } from 'react-icons/md';
import '../css/PageNotFound.css';

export default function PageNotFound() {
  return (
    <div className="PNF-container main-container">
      <div className="PNF-404">
        <MdOutlineWrongLocation className="PNF-icon" />
        <h1>4&nbsp;&nbsp;&nbsp;&nbsp;4</h1>
      </div>
      <h3>Your litter will have to be recycled elsewhere.</h3>
      <div className="PNF-buttons lower-buttons">
        <Link to="/">
          <button type="button">
            Home
          </button>
        </Link>
      </div>
    </div>
  );
}
