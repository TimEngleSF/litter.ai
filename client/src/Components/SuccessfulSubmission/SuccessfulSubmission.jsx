/* eslint-disable react/jsx-no-useless-fragment */
import React, { useState, useEffect } from 'react';
import { Link, useParams, useNavigate } from 'react-router-dom';
import { v4 as uuid } from 'uuid';
import SubmissionModal from './SubmissionModal';
import Icons from '../Icons';
import categoryData from '../../MockData/mockCategoryData';
import '../../css/SuccessfulSubmission.css';

export default function SuccessfulSubmission() {
  const { category } = useParams();
  const navigate = useNavigate();
  const [type] = useState(categoryData[category]);
  const [categoryName] = useState(category);
  const [modalOpen, setModalOpen] = useState(false);

  const separateString = (text) => (
    text.split('\n').map((line) => (
      <p key={uuid()}>{line}</p>
    ))
  );

  useEffect(() => {
    if (!categoryData[category]) {
      navigate('/404', { replace: true });
    }
  }, [category, navigate]);

  return (
    <>
      {type
      && (
      <div className="category-container main-container">
        <div className="category-wrapper">
          <h4>{type.category}</h4>
          <Icons name={categoryName} classname="category-icon" />
          <h1>{type.name}</h1>
          <div className="category-short-desc">{separateString(type.description)}</div>
          <div className="category-buttons lower-buttons">
            <button
              type="button"
              data-testid="modal-learn-more"
              onClick={() => { setModalOpen(!modalOpen); }}
            >
              Learn More
            </button>
            <Link to="/capture">
              <button type="button">
                Capture another photo
              </button>
            </Link>
            <Link to="/">
              <button className="button-home" type="button">
                Home
              </button>
            </Link>
          </div>
        </div>
        {modalOpen && (
          <>
            <div className="category-modal">
              <SubmissionModal
                modalOpen={modalOpen}
                setModalOpen={setModalOpen}
                separateString={separateString}
                type={type}
              />
            </div>
            <div
              className="modal-background"
              role="presentation"
              onClick={() => { setModalOpen(!modalOpen); }}
              onKeyDown={() => { setModalOpen(!modalOpen); }}
            />
          </>
        )}
      </div>
      )}
    </>
  );
}
