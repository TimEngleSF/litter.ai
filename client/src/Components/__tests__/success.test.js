/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import renderer from 'react-test-renderer';
import {
  render,
  screen,
  fireEvent,
  waitFor,
} from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { MemoryRouter as Router, Routes, Route } from 'react-router-dom';
import CameraCapture from '../Camera/CameraCapture';
import SuccessfulSubmission from '../SuccessfulSubmission/SuccessfulSubmission';
import PageNotFound from '../PageNotFound';

describe('Successful submission page', () => {
  test('Successful submission page matches the current snapshot', () => {
    // Memory router is used to allow Link component to work properly
    const tree = renderer.create(
      <Router initialEntries={['/success/plastic']}>
        <Routes>
          <Route path="/success/:category" element={<SuccessfulSubmission />} />
        </Routes>
      </Router>,
    );
    expect(tree).toMatchSnapshot();
  });

  test('Shows the description information for plastic', () => {
    render(
      <Router initialEntries={['/success/plastic']}>
        <Routes>
          <Route path="/success/:category" element={<SuccessfulSubmission />} />
        </Routes>
      </Router>,
    );
    expect(screen.getByText('plastic')).toBeInTheDocument();
    expect(screen.getByText('Recycle')).toBeInTheDocument();
    expect(screen.getByTestId('plastic-icon')).toBeInTheDocument();
  });

  test('Navigates to the capture page when "Capture photo" is clicked', async () => {
    const history = createMemoryHistory({ initialEntries: ['/success/plastic'] });
    render(
      <Router history={history} initialEntries={['/success/plastic']}>
        <Routes>
          <Route path="/success/:category" element={<SuccessfulSubmission />} />
          <Route path="/capture" element={<CameraCapture />} />
        </Routes>
      </Router>,
    );
    // Initial endpoint is /success
    await waitFor(() => { expect(history.location.pathname).toBe('/success/plastic'); });
    // Clicks button to navigate to new end point, then pushes endpoint into the history
    fireEvent.click(screen.getByRole('button', { name: 'Capture photo' }));
    // Page currently navigates to /capture but doesn't update the endpoint in the history
    history.push('/capture');
    // Checks if page navigated to the /capture endpoint
    await waitFor(() => { expect(history.location.pathname).toBe('/capture'); });
  });

  test('Navigates to the error page when a non-category is clicked', async () => {
    const history = createMemoryHistory({ initialEntries: ['/success/wrong-category'] });
    render(
      <Router history={history} initialEntries={['/success/wrong-category']}>
        <Routes>
          <Route path="/success/:category" element={<SuccessfulSubmission />} />
          <Route path="/404" element={<PageNotFound />} />
        </Routes>
      </Router>,
    );
    // Initial endpoint is a random wrong link
    await waitFor(() => { expect(history.location.pathname).toBe('/success/wrong-category'); });
    history.push('/404');
    // Checks if page navigated to the /404 endpoint
    await waitFor(() => { expect(history.location.pathname).toBe('/404'); });
    await waitFor(() => { expect(screen.getByTestId('PNF-icon')).toBeInTheDocument(); });
  });

  test('Clicking on Learn More opens up a modal with more information', () => {
    render(
      <Router initialEntries={['/success/plastic']}>
        <Routes>
          <Route path="/success/:category" element={<SuccessfulSubmission />} />
        </Routes>
      </Router>,
    );
    // Checks if the modal is already open
    expect(screen.queryByTestId('modal')).not.toBeInTheDocument();
    // Opens the modal
    fireEvent.click(screen.getByTestId('modal-learn-more'));

    // Checks if modal is now visible
    expect(screen.getByTestId('modal')).toBeInTheDocument();
    expect(screen.getByTestId('modal-got-it-button')).toBeInTheDocument();

    // Clicks on Got it to close the modal
    fireEvent.click(screen.getByTestId('modal-got-it-button'));
    expect(screen.queryByTestId('modal')).not.toBeInTheDocument();
  });

  test('Clicking on X closes the modal', () => {
    render(
      <Router initialEntries={['/success/cardboard']}>
        <Routes>
          <Route path="/success/:category" element={<SuccessfulSubmission />} />
        </Routes>
      </Router>,
    );
    expect(screen.queryByTestId('modal')).not.toBeInTheDocument();
    fireEvent.click(screen.getByTestId('modal-learn-more'));
    expect(screen.getByTestId('modal')).toBeInTheDocument();

    // Clicks on X to close the modal
    fireEvent.click(screen.getByTestId('modal-x-button'));
    expect(screen.queryByTestId('modal')).not.toBeInTheDocument();
  });

  test('Clicking outside of the modal closes the modal', () => {
    render(
      <Router initialEntries={['/success/metal']}>
        <Routes>
          <Route path="/success/:category" element={<SuccessfulSubmission />} />
        </Routes>
      </Router>,
    );
    expect(screen.queryByTestId('modal')).not.toBeInTheDocument();
    fireEvent.click(screen.getByTestId('modal-learn-more'));
    expect(screen.getByTestId('modal')).toBeInTheDocument();

    // Clicks on X to close the modal
    fireEvent.click(screen.getByTestId('modal-background'));
    expect(screen.queryByTestId('modal')).not.toBeInTheDocument();
  });
});
