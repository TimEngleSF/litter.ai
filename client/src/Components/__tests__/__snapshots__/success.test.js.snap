// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Successful submission page Successful submission page matches the current snapshot 1`] = `
<div
  className="category-container main-container"
>
  <div
    className="category-wrapper"
  >
    <h4>
      Recycle
    </h4>
    <svg
      className="category-icon"
      data-testid="plastic-icon"
      fill="none"
      height="1em"
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
        stroke="none"
      />
      <path
        d="M10 5h4v-2a1 1 0 0 0 -1 -1h-2a1 1 0 0 0 -1 1v2z"
      />
      <path
        d="M14 3.5c0 1.626 .507 3.212 1.45 4.537l.05 .07a8.093 8.093 0 0 1 1.5 4.694v6.199a2 2 0 0 1 -2 2h-6a2 2 0 0 1 -2 -2v-6.2c0 -1.682 .524 -3.322 1.5 -4.693l.05 -.07a7.823 7.823 0 0 0 1.45 -4.537"
      />
      <path
        d="M7 14.803a2.4 2.4 0 0 0 1 -.803a2.4 2.4 0 0 1 2 -1a2.4 2.4 0 0 1 2 1a2.4 2.4 0 0 0 2 1a2.4 2.4 0 0 0 2 -1a2.4 2.4 0 0 1 1 -.805"
      />
    </svg>
    <h1>
      plastic
    </h1>
    <div
      className="category-short-desc"
    >
      <p>
        Not all plastic is recyclable, even if you see the chasing arrows sign!
      </p>
      <p>
        If the plastic sign has a number #1 or #2 inside of the chasing arrows, it is most likely recyclable.
      </p>
    </div>
    <div
      className="category-buttons lower-buttons"
    >
      <button
        data-testid="modal-learn-more"
        onClick={[Function]}
        type="button"
      >
        Learn More
      </button>
      <a
        href="/capture"
        onClick={[Function]}
      >
        <button
          type="button"
        >
          Capture photo
        </button>
      </a>
    </div>
  </div>
</div>
`;
